Rubberduck.Parsing/Symbols/ModuleBodyElementDeclaration.cs

```diff
@@ -117,10 +73,78 @@ protected void AddParameters(IEnumerable<ParameterDeclaration> parameters)
             _parameters.AddRange(parameters);
         }
 
+        private Lazy<(bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented)> _interface;
+
+        /// <summary>
+        /// Returns true if this member is a concrete implementation of an interface.
+        /// </summary>
+        public bool IsInterfaceImplementation => _interface != null && _interface.Value.IsInterfaceImplementation;
+
+        /// <summary>
+        /// Returns the interface that this member implements from, or null if not an implementation.
+        /// </summary>
+        public ClassModuleDeclaration InterfaceImplemented => _interface?.Value.InterfaceImplemented;
+
+        private Lazy<Declaration> _implemented;
+
+        /// <summary>
+        /// Returns the interface member that this member is a concrete implementation of, or null if not an implementation.
+        /// </summary>
+        public Declaration InterfaceMemberImplemented => _implemented?.Value;
+
+        /// <inheritdoc/>
+        public string ImplementingIdentifierName => this.ImplementingIdentifierName();
+
+        /// <summary>
+        /// Returns true if this Declaration is a concrete implementation of the passed member.
+        /// </summary>
+        /// <param name="interfaceMember">The member to test for implementation of.</param>
+        /// <returns>True if this Declaration is a concrete implementation of interfaceMember.</returns>
+        protected abstract bool Implements(ICanBeInterfaceMember interfaceMember);
+
         /// <inheritdoc/>
         public bool IsInterfaceMember => this.IsInterfaceMember();
 
         /// <inheritdoc/>
         public ClassModuleDeclaration InterfaceDeclaration => this.InterfaceDeclaration();
+
+        internal void InvalidateInterfaceCache()
+        {
+            if (!(ParentDeclaration is ClassModuleDeclaration))
+            {
+                return;
+            }
+
+            _interface = new Lazy<(bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented)>(() => ResolveInterface(this), true);
+            _implemented = new Lazy<Declaration>(() => MemberImplemented(this), true);
+        }
+
+        private static (bool IsInterfaceImplementation, ClassModuleDeclaration InterfaceImplemented) ResolveInterface(ModuleBodyElementDeclaration element)
```

Comment:
this might make more sense as a Try pattern. `InterfaceImplemented` will always be `null` if the boolean is `false`:

```
private static bool TryResolveInterface(ModuleBodyElementDeclaration member, out ClassModuleDeclaration interface)
```

Makes it a bit clearer that the boolean determines what's in the second tuple part.

Reply:
Not sure I can do that with a `Lazy<T>` initializer without putting a wrapper function around it, which would probably make it less clear what's going on.